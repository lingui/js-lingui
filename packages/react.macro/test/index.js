import fs from "fs"
import glob from "glob"
import path from "path"
import { transformFileSync, transform } from "babel-core"

function getTestName(testPath) {
  return path.basename(testPath)
}

describe("macro", function() {
  const babelOptions = {
    babelrc: false,
    plugins: ["syntax-jsx", "macros"]
  }

  const transformCode = code => () => {
    try {
      transform(
        // implicitly import all lingui-react components, otherwise components
        // are ignored and not validated
        `import { 
      Trans, Plural, Select, SelectOrdinal, DateFormat, NumberFormat 
    } from '@lingui/react.macro';
    ${code}`,
        { ...babelOptions, filename: __filename } // Must pass filename otherwise babel-plugin-macros throws an error
      )
    } catch (e) {
      e.message = e.message.replace(/^.*: @lingui\/react\.macro/, "undefined") // Removes filename from the jest snapshot

      throw e
    }
  }

  const re = /[/.][^/.]/

  glob.sync(path.join(__dirname, "fixtures/*/")).forEach(testPath => {
    const testName = getTestName(testPath)
    const actualPath = path.relative(
      process.cwd(),
      path.join(testPath, "actual.js")
    )
    const expectedPath = path.join(testPath, "expected.js")

    it(testName, () => {
      let originalEnv
      if (testName.startsWith("env-production-")) {
        originalEnv = process.env.NODE_ENV
        process.env.NODE_ENV = "production"
      }

      const expected = fs.readFileSync(expectedPath, "utf8").replace(/\r/g, "")
      const actual = transformFileSync(actualPath, babelOptions)
        .code.replace(/\r/g, "")
        .trim()

      expect(actual).toEqual(expected.trim())

      if (originalEnv) process.env.NODE_ENV = originalEnv
    })
  })

  describe("validation", function() {
    describe("Plural/Select/SelectOrdinal", function() {
      it("children are not allowed", function() {
        expect(
          transformCode("<Plural>Not allowed</Plural>")
        ).toThrowErrorMatchingSnapshot()
        expect(
          transformCode("<Select>Not allowed</Select>")
        ).toThrowErrorMatchingSnapshot()
        expect(
          transformCode("<SelectOrdinal>Not allowed</SelectOrdinal>")
        ).toThrowErrorMatchingSnapshot()
      })

      it("value is missing", function() {
        const code = `<Plural one="Book" other="Books" />`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it("offset must be number or string, not variable", function() {
        const variable = `<Plural value={value} offset={offset} one="Book" other="Books" />`
        expect(transformCode(variable)).toThrowErrorMatchingSnapshot()
      })

      it("plural forms are missing", function() {
        const plural = `<Plural value={value} />`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const select = `<Select value={value} />`
        expect(transformCode(select)).toThrowErrorMatchingSnapshot()

        const ordinal = `<SelectOrdinal value={value} />`
        expect(transformCode(ordinal)).toThrowErrorMatchingSnapshot()
      })

      it("plural forms missing fallback", function() {
        const plural = `<Plural value={value} one="Book" />`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const select = `<Select value={value} one="Book" />`
        expect(transformCode(select)).toThrowErrorMatchingSnapshot()

        const ordinal = `<SelectOrdinal value={value} one="Book" />`
        expect(transformCode(ordinal)).toThrowErrorMatchingSnapshot()
      })

      it("plural rules must be valid", function() {
        const plural = `<Plural value={value} three="Invalid" one="Book" other="Books" />`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const ordinal = `<SelectOrdinal value={value} three="Invalid" one="st" other="rd" />`
        expect(transformCode(ordinal)).toThrowErrorMatchingSnapshot()
      })
    })

    describe("Date/Number", function() {
      it("value of number must be a variable", function() {
        expect(
          transformCode("<Trans><NumberFormat /></Trans>")
        ).toThrowErrorMatchingSnapshot()
      })

      it("format must be string, variable or object with custom format", function() {
        expect(
          transformCode(
            '<Trans><NumberFormat value={value} format="custom" /></Trans>'
          )
        ).not.toThrow()
        expect(
          transformCode(
            '<Trans><NumberFormat value={value} format={"custom"} /></Trans>'
          )
        ).not.toThrow()
        expect(
          transformCode(
            "<Trans><NumberFormat value={value} format={custom} /></Trans>"
          )
        ).not.toThrow()
        expect(
          transformCode(
            "<Trans><NumberFormat value={value} format={{ digits: 4 }} /></Trans>"
          )
        ).not.toThrow()
        expect(
          transformCode(
            "<Trans><NumberFormat value={value} format={42} /></Trans>"
          )
        ).toThrowErrorMatchingSnapshot()
      })

      it("value of date must be a variable", function() {
        expect(
          transformCode("<Trans><DateFormat /></Trans>")
        ).toThrowErrorMatchingSnapshot()
      })
    })
  })
})
