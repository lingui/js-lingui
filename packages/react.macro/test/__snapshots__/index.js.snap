// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macro validation Date/Number format must be string, variable or object with custom format 1`] = `undefined: Format can be either string for buil-in formats, variable or object for custom defined formats. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Date/Number value of date must be a variable 1`] = `undefined: Value argument is missing. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Date/Number value of number must be a variable 1`] = `undefined: Value argument is missing. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal children are not allowed 1`] = `undefined: Children of Plural aren't allowed. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal children are not allowed 2`] = `undefined: Children of Select aren't allowed. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal children are not allowed 3`] = `undefined: Children of SelectOrdinal aren't allowed. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal offset must be number or string, not variable 1`] = `undefined: Offset argument cannot be a variable. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal plural forms are missing 1`] = `undefined: Missing plural choices. At least fallback argument 'other' is required. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal plural forms are missing 2`] = `undefined: Missing select choices. At least fallback argument 'other' is required. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal plural forms are missing 3`] = `undefined: Missing selectordinal choices. At least fallback argument 'other' is required. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal plural forms missing fallback 1`] = `undefined: Missing fallback argument 'other'. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal plural forms missing fallback 2`] = `undefined: Missing fallback argument 'other'. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal plural forms missing fallback 3`] = `undefined: Missing fallback argument 'other'. Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal plural rules must be valid 1`] = `undefined: Invalid plural rule 'three'. Must be zero, one, two, few, many, other or exact number depending on your source language ('one' and 'other' for English). Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal plural rules must be valid 2`] = `undefined: Invalid plural rule 'three'. Must be zero, one, two, few, many, other or exact number depending on your source language ('one' and 'other' for English). Learn more: https://www.npmjs.com/package/@lingui`;

exports[`macro validation Plural/Select/SelectOrdinal value is missing 1`] = `undefined: Value argument is missing. Learn more: https://www.npmjs.com/package/@lingui`;
